pkg/virtctl/expose/expose.go

```diff
@@ -0,0 +1,186 @@
+	cmd.Flags().Int32Var(&Port, "port", 0, "The port that the service should serve on")
+	cmd.MarkFlagRequired("port")
+	cmd.Flags().StringVar(&Protocol, "protocol", "TCP", "The network protocol for the service to be created.")
+	cmd.Flags().IntVar(&TargetPort, "target-port", 0, "Name or number for the port on the VM that the service should direct traffic to. Optional.")
+	cmd.Flags().Int32Var(&NodePort, "node-port", 0, "Port used to expose the service on each node in a cluster.")
+	cmd.Flags().StringVar(&ServiceType, "type", "ClusterIP", "Type for this service: ClusterIP, NodePort, or LoadBalancer.")
+	cmd.Flags().StringVar(&PortName, "port-name", "", "Name of the port. Optional.")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+
+	return cmd
+}
+
+func usage() string {
+	usage := "  # Expose SSH to a virtual machine called 'myvm' as a node port (5555) of the cluster:\n"
+	usage += fmt.Sprintf("  virtctl expose myvm --port=5555 --target-port=22 --name=myvm-ssh --type=NodePort")
+	return usage
+}
+
+// executing the "expose" command
+func (o *Command) RunE(cmd *cobra.Command, args []string) error {
+	// first argument must be name of the VM
+	vmName := args[0]
+
+	// these are used to convert the flag values into service spec values
+	var protocol v1.Protocol
+	var targetPort intstr.IntOrString
+	var serviceType v1.ServiceType
+
+	// convert from integer to the IntOrString type
+	targetPort = intstr.FromInt(TargetPort)
+
+	// convert from string to the protocol enum
+	switch Protocol {
+	case "TCP":
+		protocol = v1.ProtocolTCP
+	case "UDP":
+		protocol = v1.ProtocolUDP
+	default:
+		return fmt.Errorf("Unknown protocol: %s", Protocol)
+	}
+
+	// convert from string to the service type enum
+	switch ServiceType {
+	case "ClusterIP":
+		serviceType = v1.ServiceTypeClusterIP
+	case "NodePort":
+		serviceType = v1.ServiceTypeNodePort
+	case "LoadBalancer":
+		serviceType = v1.ServiceTypeLoadBalancer
+	case "ExternalName":
+		return fmt.Errorf("Type: %s not supported", ServiceType)
+	default:
+		return fmt.Errorf("Unknown service type: %s", ServiceType)
+	}
+
+	// get the namespace
+	namespace, _, err := o.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	// get the client
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(o.clientConfig)
+	if err != nil {
+		return fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
+	}
+
+	// TODO: not sure what are the options used for?
+	options := k8smetav1.GetOptions{}
+	// get the VM
+	_, vm_err := virtClient.VM(namespace).Get(vmName, options)
+	if vm_err != nil {
+		// try to get offline VM this is only for better error message
+		_, ovm_err := virtClient.OfflineVirtualMachine(namespace).Get(vmName, &options)
+		if ovm_err != nil {
+			return fmt.Errorf("Error fetching VirtualMachine: %v", vm_err)
+		}
+		return fmt.Errorf("OfflineVirtualMachine: %s must be started before being exposed", vmName)
```

Comment:
To be declarative, we should not assume that it has to be started. We basically define: "if vm is started start routing traffic through this service", when looking at it from the offline virtual machine perspective.

Reply:
ok. will remove (also test this scenario)